### PR TITLE
Compression loss device bug

### PR DESCRIPTION
**Pull request by SKholkin**
ticket: 49963

### CODE REVIEW

File: nncf/compression_method_api.py [hunk 1]
```diff
@@ -51,14 +51,20 @@ class PTCompressionLoss(nn.Module, CompressionLoss):
     sparsity algorithm calculates the number of non-zero weights in convolutional
     and fully-connected layers to construct the loss function.
     """
+    def __init__(self, default_device=None):
+        """
+        :param default_device: torch device being used for computing internal tensors
```
Comment (ljaljushkin) [inline nncf/compression_method_api.py:56]: it's not actually for computing, it's just for creating a default value of loss on the target device

File: tests/test_algo_common.py [hunk 2]
```diff
@@ -381,3 +382,60 @@ def test_can_export_compressed_model_with_specified_domain_for_custom_ops(tmp_pa
...
+                        "sparsity_target_epoch": 2,
+                        "sparsity_freeze_epoch": 3
+                    },
+            }
+        ]
+    })
+                      ]
+
+
+@pytest.mark.parametrize("is_cpu", [True, False], ids=['is_cpu_' + 'True', 'is_cpu_' + 'False'])
```
Comment (ljaljushkin) [inline tests/test_algo_common.py:431]: cpu and cuda is more meaningful

File: tests/test_algo_common.py [hunk 3]
```diff
@@ -381,3 +382,60 @@ def test_can_export_compressed_model_with_specified_domain_for_custom_ops(tmp_pa
...
+
+def get_updated_config(config, update):
+    config.update(update)
+    return config
+
+
+comp_loss_configs = [get_updated_config(get_basic_sparsity_config(), {
+    "compression": [
+        {
+            "algorithm": "rb_sparsity",
```
Comment (ljaljushkin) [inline tests/test_algo_common.py:395]: please remove duplication of code by defining common parts

File: examples/classification/configs/quantization/mobilenet_v2_imagenet_int8.json [hunk 4]
```diff
@@ -15,12 +15,13 @@
...
+    "compression": [
+        {
         "algorithm": "quantization",
         "initializer": {
             "range": {
                 "num_init_samples": 2560
             }
         }
-    }
+        }]
```
Comment (ljaljushkin) [inline examples/classification/configs/quantization/mobilenet_v2_imagenet_int8.json:26]: no need to change it, please revert

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/compression_method_api.py [hunk 5]
```diff
@@ -51,14 +51,20 @@ class PTCompressionLoss(nn.Module, CompressionLoss):
     sparsity algorithm calculates the number of non-zero weights in convolutional
     and fully-connected layers to construct the loss function.
     """
+    def __init__(self, default_device=None):
+        """
+        :param default_device: torch device being used for creating default value loss tensor
```
Comment (ljaljushkin) [inline nncf/compression_method_api.py:56]: for creating a default value of loss on the target device

File: tests/quantization/test_scheduler.py [hunk 6]
```diff
@@ -33,7 +34,9 @@ def create_staged_scheduler(ctrl_spy, w_start=2, a_start=1):
 class QuantizationCtrlBaseSpy:
     #pylint:disable=no-member
     def __init__(self, mocker):
-        self._mocked_ctrl = QuantizationControllerBase(mocker.stub)
+        mock_model = mocker.MagicMock(spec=NNCFNetwork)
```
Comment (ljaljushkin) [inline tests/quantization/test_scheduler.py:37]: 👍

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: tests/test_algo_common.py [hunk 7]
```diff
@@ -381,3 +385,58 @@ def test_can_export_compressed_model_with_specified_domain_for_custom_ops(tmp_pa
             count_custom_ops += 1
 
     assert count_custom_ops == 4
+
+
+def charnge_compression_algorithms_order(config):
```
Comment (ljaljushkin) [inline tests/test_algo_common.py:390]: please be more attentive! 
charnge -> change

File: nncf/compression_method_api.py [hunk 8]
```diff
@@ -51,14 +51,20 @@ class PTCompressionLoss(nn.Module, CompressionLoss):
     sparsity algorithm calculates the number of non-zero weights in convolutional
     and fully-connected layers to construct the loss function.
     """
+    def __init__(self, default_device=None):
+        """
+        :param default_device: torch device being used for creating a default value of loss on the target device
```
Comment (vshampor) [inline nncf/compression_method_api.py:56]: I am fairly sure that everywhere we have to fall back on the default value of the loss we also know the device of the model to be compressed. I suggest that instead of changing the API for this you keep the API as-is and instead define a new `ZeroCompressionLoss(PTCompressionLoss)` with an init function taking a device. Instantiate it explicitly in all algos that do not define their specific compression loss.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/compression_method_api.py (hunk 8 above)
Comment (ljaljushkin) [reply nncf/compression_method_api.py:56]: With current design I see the problem with implementation of PTCompressionAlgorithmController:
```
def __init__(self, target_model: ModelType):
     self._loss = PTCompressionLoss()
```
Algorithms are not required to define a new ZeroCompressionLoss. That still would lead to errors with device.

Comment (vshampor) [reply nncf/compression_method_api.py:56]: Even the `PTCompressionAlgorithmController` should be an interface, not a real base class. The real controller class specifying no compression is `NoCompressionAlgorithmController`.

Comment (ljaljushkin) [reply nncf/compression_method_api.py:56]: it means that we need to change current design first (probably in a separate PR), because this class has a lot of implemented methods: `export_model`, `run_batchnorm_adaptation`, `statistics`

File: tests/test_algo_common.py (hunk 7 above)
Comment (vshampor) [reply tests/test_algo_common.py:388]: I can still see `charnge` here.

**Review by vshampor — CHANGES_REQUESTED** [review]

**Comment by vshampor** [comment]
Jenkins please retry a build

File: nncf/compression_method_api.py (hunk 8 above)
Comment (vshampor) [reply nncf/compression_method_api.py]: Will be done in a separate PR.

**Review by vshampor — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
